testing/src/main/scala/prometheus4cats/testing/TestingMetricRegistry.scala

```diff
@@ -0,0 +1,426 @@
+      labelNames: IndexedSeq[Label.Name],
+      quantiles: Seq[Summary.QuantileDefinition],
+      maxAge: FiniteDuration,
+      ageBuckets: Summary.AgeBuckets
+  )(f: A => IndexedSeq[String]): Resource[F, Summary.Labelled[F, Double, A]] =
+    store(
+      prefixedName(prefix, name.value),
+      names(commonLabels, labelNames),
+      MetricType.Summary,
+      (ref: MapRef[F, List[String], Chain[Double]]) =>
+        Summary.Labelled.make[F, Double, A]((d: Double, a: A) => ref(values(commonLabels, f(a))).update(_.append(d))),
+      Chain.nil
+    )
+
+  override protected[prometheus4cats] def createAndRegisterInfo(
+      prefix: Option[Metric.Prefix],
+      name: Info.Name,
+      help: Metric.Help
+  ): Resource[F, Info[F, Map[Label.Name, String]]] = {
+    val release = info(prefixedName(prefix, name.value)).update {
+      case None => throw new RuntimeException("This should be unreachable - our reference counting has a bug")
+      case Some((n, i)) => if (n == 1) None else Some(n - 1 -> i)
+
+    }
+    Resource.make(
+      info(prefixedName(prefix, name.value)).modify {
+        case None =>
+          val i = Info.make[F, Map[Label.Name, String]]((_: Map[Label.Name, String]) => F.unit)
+          Some(1 -> i) -> i
+        case Some((n, i)) => Some(n + 1 -> i) -> i
+      }
+    )(_ => release)
+  }
+
+  private def store[M <: Metric[Double]](
+      name: String,
+      labels: List[String],
+      tpe: MetricType,
+      create: MapRef[F, List[String], Chain[Double]] => M,
+      initial: Chain[Double]
+  ): Resource[F, M] =
+    Resource
+      .eval(
+        MapRef
+          .ofShardedImmutableMap[F, List[String], Chain[Double]](32)
+          .map(r => MapRef.defaultedMapRef[F, List[String], Chain[Double]](r, initial))
+          .flatMap { ref =>
+            val release =
+              underlying(name -> labels).update {
+                case None => throw new RuntimeException("This should be unreachable - our reference counting has a bug")
+                case Some((n, t, c, h)) => if (n == 1) None else Some((n - 1, t, c, h))
+              }
+
+            underlying(name -> labels).modify {
+              case None =>
+                val m = create(ref)
+                Some((1, tpe, m, ref)) -> F.pure(
+                  Resource.make(F.pure(m))(_ => release)
+                )
+              case curr @ Some((n, t, m, h)) =>
+                if (t == tpe)
+                  Some((n + 1, t, m, h)) ->
+                    F.pure(
+                      Resource.make(
+                        // Cast safe by construction
+                        F.pure(m.asInstanceOf[M])
+                      )(_ => release)
+                    )
+                else
+                  curr -> F.raiseError[Resource[F, M]](
+                    new RuntimeException(
+                      s"Cannot create metric of type $tpe as metric of type $t alreasy exists with the same name and labels"
```

Comment:
In the long term we want to support the ability to have a metric with the same name and different labels. I'm wondering if we do that here with the testing impl or wait until we have a better alternative.

Reply:
Also typo in `alreasy`

Reply:
Ann I missing something? 😅 What is wrong with the message sorry?

Reply:
Oh no, it's me 🤦. What you have is correct!